exchanges/coinbene/coinbene_websocket.go

```diff
@@ -0,0 +1,321 @@
+	for x := range channels {
+		for y := range c.EnabledPairs {
+			c.EnabledPairs[y].Delimiter = ""
+			subscriptions = append(subscriptions, wshandler.WebsocketChannelSubscription{
+				Channel:  fmt.Sprintf(channels[x], c.EnabledPairs[y]),
+				Currency: c.EnabledPairs[y],
+			})
+		}
+	}
+	c.Websocket.SubscribeToChannels(subscriptions)
+}
+
+// GenerateAuthSubs generates auth subs
+func (c *Coinbene) GenerateAuthSubs() {
+	var subscriptions []wshandler.WebsocketChannelSubscription
+	var sub wshandler.WebsocketChannelSubscription
+	var userChannels = []string{"user.account", "user.position", "user.order"}
+	for z := range userChannels {
+		sub.Channel = userChannels[z]
+		subscriptions = append(subscriptions, sub)
+	}
+	c.Websocket.SubscribeToChannels(subscriptions)
+}
+
+// WsDataHandler handles websocket data
+func (c *Coinbene) WsDataHandler() {
+	c.Websocket.Wg.Add(1)
+
+	defer c.Websocket.Wg.Done()
+
+	for {
+		select {
+		case <-c.Websocket.ShutdownC:
+			return
+
+		default:
+			stream, err := c.WebsocketConn.ReadMessage()
+			if err != nil {
+				c.Websocket.DataHandler <- err
+				return
+			}
+			c.Websocket.TrafficAlert <- struct{}{}
+			if string(stream.Raw) == "ping" {
+				c.WebsocketConn.Lock()
+				c.WebsocketConn.Connection.WriteMessage(websocket.TextMessage, []byte("pong"))
+				c.WebsocketConn.Unlock()
+				continue
+			}
+			var result map[string]interface{}
+			err = common.JSONDecode(stream.Raw, &result)
+			if err != nil {
+				c.Websocket.DataHandler <- err
+			}
+			_, ok := result["event"]
+			switch {
+			case ok && result["event"].(string) == "subscribe":
+				continue
+			case ok && result["event"].(string) == "unsubscribe":
+				continue
+			case ok && result["event"].(string) == "error":
+				c.Websocket.DataHandler <- fmt.Errorf("message: %s. code: %v", result["message"], result["code"])
+				continue
+			}
+			if ok && strings.Contains(result["event"].(string), "login") {
+				if result["success"].(bool) {
+					c.GenerateAuthSubs()
+				}
+				continue
+			}
+			switch {
+			case strings.Contains(result["topic"].(string), "ticker"):
+				var ticker WsTicker
+				err = common.JSONDecode(stream.Raw, &ticker)
+				if err != nil {
+					c.Websocket.DataHandler <- err
+					continue
+				}
+				for x := range ticker.Data {
+					c.Websocket.DataHandler <- wshandler.TickerData{
+						Quantity:   ticker.Data[x].Volume24h,
+						ClosePrice: ticker.Data[x].LastPrice,
+						HighPrice:  ticker.Data[x].High24h,
+						LowPrice:   ticker.Data[x].Low24h,
+						Pair:       currency.NewPairFromString(ticker.Data[x].Symbol),
+						Exchange:   c.Name,
+						AssetType:  orderbook.Swap,
+					}
+				}
+			case strings.Contains(result["topic"].(string), "tradeList"):
+				var tradeList WsTradeList
+				err = common.JSONDecode(stream.Raw, &tradeList)
+				if err != nil {
+					c.Websocket.DataHandler <- err
+					continue
+				}
+				time, err1 := time.Parse(time.RFC3339, tradeList.Data[0][3])
```

Comment:
For neatness, can initalisate var t time.Time and reuse err above instead of creating a new err per func call